ROCm: TypeError: operand type(s) all returned NotImplemented from __array_ufunc__

Previously all FFT tests passed on HIP, but it seems the recent change on `__array_ufunc__` (#4198) somehow breaks it...
```
$ pytest tests/cupy_tests/fft_tests/
(...similar error output omitted...)
__________________________________________________________ TestFftfreq.test_rfftfreq[_param_2_{d=2, n=100}] ___________________________________________________________

self = <<cupy_tests.fft_tests.test_fft.TestFftfreq object at 0x7f4144ceb090>  parameter: {'n': 100, 'd': 2}>
xp = <module 'cupy' from '/home/leofang/dev/cupy_rocm350/cupy/__init__.py'>, dtype = <class 'numpy.float64'>

    @testing.for_all_dtypes()
    @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, contiguous_check=False)
    def test_rfftfreq(self, xp, dtype):
>       out = xp.fft.rfftfreq(self.n, self.d)

tests/cupy_tests/fft_tests/test_fft.py:1258: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cupy/fft/_fft.py:1017: in rfftfreq
    return cupy.arange(0, n // 2 + 1, dtype=np.float64) / (n * d)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   return numpy.true_divide(x, y)
E   AssertionError: Only cupy raises error
E   
E   Traceback (most recent call last):
E     File "/home/leofang/dev/cupy_rocm350/cupy/testing/helper.py", line 47, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/leofang/dev/cupy_rocm350/tests/cupy_tests/fft_tests/test_fft.py", line 1258, in test_rfftfreq
E       out = xp.fft.rfftfreq(self.n, self.d)
E     File "/home/leofang/dev/cupy_rocm350/cupy/fft/_fft.py", line 1017, in rfftfreq
E       return cupy.arange(0, n // 2 + 1, dtype=np.float64) / (n * d)
E     File "cupy/core/core.pyx", line 1103, in cupy.core.core.ndarray.__truediv__
E       return numpy.true_divide(x, y)
E   TypeError: operand type(s) all returned NotImplemented from __array_ufunc__(<ufunc 'true_divide'>, '__call__', array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.,
E          13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
E          26., 27., 28., 29., 30., 31., 32., 33., 34., 35., 36., 37., 38.,
E          39., 40., 41., 42., 43., 44., 45., 46., 47., 48., 49., 50.]), 200): 'ndarray', 'int'

cupy/core/core.pyx:1103: AssertionError
------------------------------------------------------------------------ Captured stdout call -------------------------------------------------------------------------
dtype is <class 'numpy.float64'>
======================================================================= short test summary info =======================================================================
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_0_{n=None, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_1_{n=None, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_4_{n=5, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_5_{n=5, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_8_{n=10, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_9_{n=10, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_12_{n=15, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_hfft[_param_13_{n=15, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_0_{n=None, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_1_{n=None, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_2_{n=None, norm='ortho', shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_3_{n=None, norm='ortho', shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_4_{n=5, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_5_{n=5, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_6_{n=5, norm='ortho', shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_7_{n=5, norm='ortho', shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_8_{n=10, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_9_{n=10, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_10_{n=10, norm='ortho', shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_11_{n=10, norm='ortho', shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_12_{n=15, norm=None, shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_13_{n=15, norm=None, shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_14_{n=15, norm='ortho', shape=(10,)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestHfft::test_ihfft[_param_15_{n=15, norm='ortho', shape=(10, 10)}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestFftfreq::test_fftfreq[_param_0_{d=1, n=1}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestFftfreq::test_fftfreq[_param_1_{d=0.5, n=10}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestFftfreq::test_fftfreq[_param_2_{d=2, n=100}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestFftfreq::test_rfftfreq[_param_0_{d=1, n=1}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestFftfreq::test_rfftfreq[_param_1_{d=0.5, n=10}] - AssertionError: Only cupy raises error
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestFftfreq::test_rfftfreq[_param_2_{d=2, n=100}] - AssertionError: Only cupy raises error
======================================================= 30 failed, 557 passed, 727 skipped in 103.58s (0:01:43) =======================================================
```

--- a/cupy/core/core.pyx
+++ b/cupy/core/core.pyx
@@ -180,7 +180,7 @@ cdef class ndarray:
     @property
     def __cuda_array_interface__(self):
         if runtime._is_hip_environment:
-            raise RuntimeError(
+            raise AttributeError(
                 'HIP/ROCm does not support cuda array interface')
         cdef dict desc = {
             'shape': self.shape,
@@ -1439,11 +1439,12 @@ cdef class ndarray:
                 # implicit host-to-device conversion.
                 # Except for numpy.ndarray, types should be supported by
                 # `_kernel._preprocess_args`.
-                if (
-                        not hasattr(x, '__cuda_array_interface__')
+                check = hasattr(x, '__cuda_array_interface__')
+                if runtime._is_hip_environment and isinstance(x, ndarray):
+                    check = True
+                if (not check
                         and not type(x) in _scalar.scalar_type_set
-                        and not isinstance(x, numpy.ndarray)
-                ):
+                        and not isinstance(x, numpy.ndarray)):
                     return NotImplemented
             if name in [
                     'greater', 'greater_equal', 'less', 'less_equal',

--- a/tests/cupy_tests/core_tests/test_ndarray.py
+++ b/tests/cupy_tests/core_tests/test_ndarray.py
@@ -315,6 +315,22 @@ class TestNdarrayCudaInterfaceStream(unittest.TestCase):
             assert iface['stream'] == 1 if stream.ptr == 0 else stream.ptr
 
 
+@pytest.mark.skipif(not cupy.cuda.runtime.is_hip,
+                    reason='This is supported on CUDA')
+class TestNdarrayCudaInterfaceNoneCUDA(unittest.TestCase):
+
+    def setUp(self):
+        self.arr = cupy.zeros(shape=(2, 3), dtype=cupy.float64)
+
+    def test_cuda_array_interface_hasattr(self):
+        assert not hasattr(self.arr, '__cuda_array_interface__')
+
+    def test_cuda_array_interface_getattr(self):
+        with pytest.raises(AttributeError) as e:
+            getattr(self.arr, '__cuda_array_interface__')
+        assert 'HIP' in str(e.value)
+
+
 @testing.parameterize(
     *testing.product({
         'indices_shape': [(2,), (2, 3)],

